File: client/src/app/family/family.ts
// student information interface
export interface StudentInfo {
  name: string;
  grade: string;
  school: string;
  schoolAbbreviation: string;
  teacher: string;
  headphones: boolean;
  backpack: boolean;
}

// family information interface, apart of the general family information
export interface AvailabilityOptions {
  earlyMorning: boolean;
  lateMorning: boolean;
  earlyAfternoon: boolean;
  lateAfternoon: boolean;
}

export interface ChecklistItem {
  id: string;
  label: string;
  selected: boolean;
  available: boolean;
  itemDescription?: string;
  supplyListId?: string;
  matchedInventoryId?: string;
  matchedInventoryItem?: string;
  matchedInventoryDescription?: string;
  requestedQuantity: number;
  notPickedUpReason?: string;
  substituteItem?: string;
  substituteBarcode?: string;
  substituteDescription?: string;
  substituteInventoryId?: string;
  notes?: string;
}

export interface ChecklistSection {
  id: string;
  title: string;
  printableTitle: string;
  saved: boolean;
  items: ChecklistItem[];
}

export interface FamilyChecklist {
  templateId: string;
  printableTitle: string;
  snapshot: boolean;
  sections: ChecklistSection[];
}

export type ScheduleColumnType = 'English' | 'Spanish';

export interface ScheduleAssignment {
  timeSlot: string;
  columnType: ScheduleColumnType;
  columnIndex: number;
}

// main family interface, which includes general family information and an array of students in the family
export interface Family {
  _id?: string;
  ownerUserId?: string;
  profileComplete?: boolean;
  guardianName: string;
  email: string;
  address: string;
  accommodations: string;
  needSpanishHelp: boolean;
  timeSlot: string;
  scheduleAssignment?: ScheduleAssignment;
  timeAvailability?: AvailabilityOptions;
  helped?: boolean;
  status?: string;
  checklist?: FamilyChecklist | null;
  students: StudentInfo[];
  deleteRequest?: FamilyDeleteRequest;
}

export interface FamilyDeleteRequest {
  requested: boolean;
  message?: string;
  requestedByUserId?: string;
  requestedByUserName?: string;
  requestedBySystemRole?: string;
  requestedAt?: string;
}

// dashboard statistics interface
export interface DashboardStats {
  studentsPerSchool: { [school: string]: number};
  studentsPerGrade: { [grade: string]: number};
  totalFamilies: number;
  totalStudents: number;
}

// select option interface for dropdown filters
export interface SelectOption {
  label: string;
  value: string;
}

File: server/src/main/java/umm3601/Family/Family.java
// Packages
package umm3601.Family;

// Java Imports
import java.util.ArrayList;
import java.util.List;

// Org Imports
import org.mongojack.Id;
import org.mongojack.ObjectId;

@SuppressWarnings({"VisibilityModifier"})
public class Family {
  @ObjectId @Id
  @SuppressWarnings({"MemberName"})
  public String _id;

  public String ownerUserId; // Guardian-level information (applies to the whole household)
  public String guardianName;
  public String email;
  public String address;
  // Whether the guardian or volunteer has completed the family profile.
  public boolean profileComplete;
  public String timeSlot;
  public String accommodations;
  public boolean helped;
  public String status;
  public FamilyChecklist checklist;

  public List<StudentInfo> students;

  public AvailabilityOptions timeAvailability;

  public static class AvailabilityOptions {
    public boolean earlyMorning;
    public boolean lateMorning;
    public boolean earlyAfternoon;
    public boolean lateAfternoon;

    // helper function for sorting
    public int countTrue() {
      int count = 0;
      if (earlyMorning) {
        count++;
      }
      if (lateMorning) {
        count++;
      }
      if (earlyAfternoon) {
        count++;
      }
      if (lateAfternoon) {
        count++;
      }
      return count;
    }
  }

  // Optional delete request metadata used by volunteer->admin approval flow.
  public DeleteRequest deleteRequest;

  public static class StudentInfo {
    public String name;
    public String grade;
    public String school;
    public String schoolAbbreviation;
    public String teacher;
    public Boolean backpack;
    public Boolean headphones;
  }

  public static class DeleteRequest {
    public boolean requested;
    public String message;
    public String requestedByUserId;
    public String requestedByUserName;
    public String requestedBySystemRole;
    public String requestedAt;
  }

  @SuppressWarnings({"VisibilityModifier"})
  public static class FamilyChecklist {
    public String templateId;
    public String printableTitle;
    public boolean snapshot;
    public List<ChecklistSection> sections = new ArrayList<>();
  }

  @SuppressWarnings({"VisibilityModifier"})
  public static class ChecklistSection {
    public String id;
    public String title;
    public String printableTitle;
    public boolean saved;
    public List<ChecklistItem> items = new ArrayList<>();
  }

  @SuppressWarnings({"VisibilityModifier"})
  public static class ChecklistItem {
    public String id;
    public String label;
    public boolean selected = true;
    public boolean available = true;
    public String itemDescription;
    public String supplyListId;
    public String matchedInventoryId;
    public String matchedInventoryItem;
    public String matchedInventoryDescription;
    public Integer requestedQuantity = 1;
    public String notPickedUpReason;
    public String substituteItem;
    public String substituteBarcode;
    public String substituteDescription;
    public String substituteInventoryId;
    public String notes;
  }
}
